Source/ARTTypes.m

```diff
@@ -308,3 +308,74 @@ + (NSDate *)dateWithMillisecondsSince1970:(uint64_t)msecs {
 }
 
 @end
+
+@interface ARTCancellableFromCallback : NSObject<ARTCancellable>
++(instancetype)new NS_UNAVAILABLE;
+-(instancetype)init NS_UNAVAILABLE;
+-(instancetype)initWithCallback:(ARTCompletionHandler)callback;
+@property(nonatomic, readonly) ARTCompletionHandler wrapper;
+@end
+
+@implementation ARTCancellableFromCallback {
+    id _lock;
+    
+    // _callback will be nil once either cancel or invoke has been called on
+    // this instance.
+    volatile ARTCompletionHandler _callback;
+}
+
+@synthesize wrapper = _wrapper;
+
+-(instancetype)initWithCallback:(const ARTCompletionHandler)callback {
+    if (!callback) {
+        [NSException raise:NSInternalInconsistencyException format:@"callback is nil."];
+    }
+    self = [super init];
+    if (!self) {
+        return nil;
+    }
+    
+    _lock = [NSObject new];
+    _callback = callback;
+    
+    __weak ARTCancellableFromCallback * _cancellable = self;
```

Comment:
Same, why is this weak?

Reply:
This one is more important and I believe justified. We don't want `_wrapper` retaining a strong reference to `self`, given it's a strong ivar on `self` itself. Unless I missed something?